src/Rule/CompareHandler.php

```diff
@@ -30,70 +35,196 @@ public function validate(mixed $value, object $rule, ValidationContext $context)
+    }
+
+    /**
+     * Checks whether the validated value is allowed for types that require type casting - {@see CompareType::NUMBER}
+     * and {@see CompareType::STRING}.
+     *
+     * @param mixed $value The Validated value.
+     *
+     * @return bool `true` if value is allowed and `false` otherwise.
+     */
+    private function isValueAllowedForTypeCasting(mixed $value): bool
+    {
+        return $value === null ||
+            is_scalar($value) ||
+            $value instanceof Stringable ||
+            $value instanceof DateTimeInterface;
+    }
+
+    /**
+     * Gets representation of the value for using with error parameter.
+     *
+     * @param string $type The type of the values being compared ({@see AbstractCompare::$type}).
+     * @psalm-param CompareType::ORIGINAL | CompareType::STRING | CompareType::NUMBER $type
+     *
+     * @param mixed $value The мalidated value.
+     *
+     * @return scalar|null Formatted value.
+     */
+    private function getFormattedValue(string $type, mixed $value): int|float|string|bool|null
+    {
+        if ($value === null || is_scalar($value)) {
+            return $value;
+        }
+
+        if ($value instanceof Stringable && $type !== CompareType::ORIGINAL) {
+            return (string) $value;
+        }
+
+        return get_debug_type($value);
+    }
+
+    /**
+     * Compares two values according to the specified type and operator.
      *
      * @param string $operator The comparison operator. One of `==`, `===`, `!=`, `!==`, `>`, `>=`, `<`, `<=`.
-     * @param string $type The type of the values being compared.
-     * @psalm-param CompareType::STRING | CompareType::NUMBER $type
+     * @param string $type The type of the values being compared ({@see AbstractCompare::$type}).
+     * @psalm-param CompareType::ORIGINAL | CompareType::STRING | CompareType::NUMBER $type
      *
-     * @param mixed $value The value being compared.
-     * @param mixed $targetValue Another value being compared.
+     * @param mixed $value The validated value.
+     * @param mixed $targetValue "Target" value set in rule options.
      *
      * @return bool Whether the result of comparison using the specified operator is true.
      */
-    private function compareValues(string $operator, string $type, mixed $value, mixed $targetValue): bool
+    private function compareValues(string $type, string $operator, mixed $value, mixed $targetValue): bool
     {
-        if ($type === CompareType::NUMBER) {
-            $value = (float) $value;
-            $targetValue = (float) $targetValue;
-        } else {
-            $value = (string) $value;
-            $targetValue = (string) $targetValue;
+        if (!in_array($operator, ['==', '===', '!=', '!=='])) {
+            if ($type === CompareType::STRING) {
+                $value = (string) $value;
+                $targetValue = (string) $targetValue;
```

Comment:
```php
'target value: Unix Timestamp string, value: DateTime object with the same value, type: string, operator: ==' => [
    $dateTime->format('U'),
    [new Compare($dateTime, type: CompareType::STRING)],
],
```

Result: `Error: Object of class DateTime could not be converted to string`.

Into `checkValuesAreEqual()` also.